### PR TITLE
Instrument, Files, and Custom representation

### PR DESCRIPTION
**Pull request by aburrell**
# Description
Improves Class object representation for Instrument, Files, and Custom.

Addresses #465

Please include a summary of the change and which issue is fixed. Please also
include relevant motivation and context. List any dependencies that are required
for this change.  Please see ``CONTRIBUTING.md`` for more guidelines.

## Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# How Has This Been Tested?

Load a pysat Instrument object and try `print(inst)` and just `inst`.  Also, try `print(inst.files)` and `print(inst.custom)` (with and without print statements).  See the lovely object summaries!

**Test Configuration**:
* Operating system: OS X Mojave
* Version number: 3.7

# Checklist:

- [x] Make sure you are merging into the ``develop`` (not ``main``) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Add a note to ``CHANGELOG.md``, summarizing the changes

### CODE REVIEW

**Comment by aburrell** [comment]
This may require an update to the documentation, but I don't want to make changes unless #486 is done.

**Comment by aburrell** [comment]
The downstream changes aren't checked because I am not sure if the downstream docs use the inst printed output.  But I think they're mostly non-existent at this point?

**Comment by rstoneback** [comment]
> This may require an update to the documentation, but I don't want to make changes unless #486 is done.

#524 which addresses #486 is ready for a review. Cheers.

File: pysat/_instrument.py [hunk 1]
```diff
@@ -1053,32 +1053,34 @@ def _assign_funcs(self, by_name=False, inst_module=None):
...
-
-        output_str = '\npysat Instrument object\n'
+    def __repr__(self):
+        # Print the basic Instrument properties
+        output_str = 'pysat Instrument object\n'
         output_str += '-----------------------\n'
         output_str += 'Platform: ' + self.platform + '\n'
         output_str += 'Name: ' + self.name + '\n'
         output_str += 'Tag: ' + self.tag + '\n'
         output_str += 'Satellite id: ' + self.sat_id + '\n'
```
Comment (rstoneback) [inline pysat/_instrument.py:1063]:
```suggestion
    def __repr__(self):
        # return a string representation that could be used to recreate object
        # https://stackoverflow.com/questions/1436703/difference-between-str-and-repr
        output_str = 'pysat.Instrument('
        output_str += '%r, %r' % (self.platform, self.name)
        output_str += ', sat_id=%r, tag=%r)' % (self.sat_id, self.tag)
```

Comment (rstoneback) [reply pysat/_instrument.py:1063]:
```
In [1]: import pysat                                                                                                                                                       

In [2]: test = pysat.Instrument('pysat', 'testing')                                                                                                                        

In [3]: test                                                                                                                                                               
Out[3]: pysat.Instrument('pysat', 'testing', sat_id='', tag='')

```

Comment (rstoneback) [reply pysat/_instrument.py:1063]: This is lacking custom keywords and custom methods.

File: pysat/_instrument.py [hunk 2]
```diff
@@ -1053,32 +1053,34 @@ def _assign_funcs(self, by_name=False, inst_module=None):
...
         output_str += 'Cleaning Level: ' + self.clean_level + '\n'
         output_str += 'Data Padding: ' + self.pad.__repr__() + '\n'
         output_str += 'Keyword Arguments Passed to load(): '
-        output_str += self.kwargs.__repr__() + '\nCustom Functions : \n'
-        if len(self.custom._functions) > 0:
-            for func in self.custom._functions:
-                output_str += '    ' + func.__repr__() + '\n'
-        else:
-            output_str += '    ' + 'No functions applied.\n'
+        output_str += self.kwargs.__repr__() + '\n'
```
Comment (rstoneback) [inline pysat/_instrument.py:1077]:
```suggestion
        output_str += self.kwargs.__str__() + '\n'
```

File: pysat/_instrument.py [hunk 3]
```diff
@@ -1053,32 +1053,34 @@ def _assign_funcs(self, by_name=False, inst_module=None):
...
+        output_str += self.custom.__repr__()
```
Comment (rstoneback) [inline pysat/_instrument.py:1078]:
```suggestion
        output_str += self.custom.__str__()
```

Comment (rstoneback) [reply pysat/_instrument.py:1078]: and similar.

File: pysat/_instrument.py (hunk 2 above)
Comment (rstoneback) [inline pysat/_instrument.py:1077]:
```
pysat.Instrument('pysat', 'testing', sat_id='', tag='')
Data Processing
---------------
Cleaning Level: clean
Data Padding: None
Keyword Arguments Passed to load(): {'sim_multi_file_right': False, 'sim_multi_file_left': False, 'root_date': None, 'file_date_range': None, 'malformed_index': False, 'mangle_file_dates': False}
Custom Functions: 0 applied
Local File Statistics
---------------------
Number of files: 1096
Date Range: 01 January 2008 --- 31 December 2010

Loaded Data Statistics
----------------------
No loaded data.

```

It would be nice if the keyword arguments had some additional formatting.

Comment (rstoneback) [reply pysat/_instrument.py:1077]: line breaks in between sections?

Comment (aburrell) [reply pysat/_instrument.py]: I agree about the line break between sections, but not to more formatting for the keyword arguments.  I don't really see the value in printing these out here (most of them are default and not user-specified), and would prefer that they were removed.

Comment (aburrell) [reply pysat/_instrument.py]: Ok, I decided the load kwargs are ok, but I think the current formatting is fine.

File: pysat/_instrument.py (hunk 1 above)
Comment (aburrell) [reply pysat/_instrument.py]: Ok, I don't like adding the `kwargs` to `__repr__` because it contains more than just what the user specified.  However, including it is less harmful than excluding it.  Adding `custom` here is too much.  The user can easily get that info by typing `inst.custom`.  There's no short way of including all the possible ways someone has altered `inst` after the initial call.  So I've limited `__repr__` to only include things that can be specified during the initialization.

**Comment by aburrell** [comment]
Ok, looking over the docs here and elsewhere, there's nothing immediate that needs to be fixed.  However, I'd now recommend using the Instrument, Custom, and File str/repr functions in future tutorial examples.

**Comment by aburrell** [comment]
Examples and style logic:

`__repr__`:
```
import pysat
inst = pysat.Instrument('jro', 'isr', 'drifts')
inst
Instrument(platform='jro', name='isr', sat_id='', clean_level='clean', pad=None, orbit_info={'index': None, 'kind': None, 'period': None}, **{})
```

All of the saved instrument inputs are provided.  Missing the `update_files` one, because it isn't saved.  Not providing any information for the other Classes that are attached to Instrument.

```
inst.custom
Custom -> 0 functions applied
```
Custom is tricky because EVERYTHING in it is hidden.  Decided to provide a short summary of the contents.

```
inst.files
Files(Instrument(platform='jro', name='isr', sat_id='', clean_level='clean', pad=None, orbit_info={'index': None, 'kind': None, 'period': None}, **{}), manual_org=False, directory_format={platform}/{name}/{tag}, update_files=None, file_format=None, write_to_diskTrue, ignore_empty_files=False) -> 95 Local files
```
Now you see a combination of the format of Instrument and Custom.  There are visible input parameters, all of which are specified here.  Also, a short summary of the Class contents is provided.  Because `inst` is an input to Files, it's `__repr__` function is included as the first argument.

`__str__`

```
print(inst, '\n', inst.custom, '\n', inst.files)

pysat Instrument object
-----------------------
Platform: 'jro'
Name: 'isr'
Tag: 'drifts'
Satellite id: ''

Data Processing
---------------
Cleaning Level: 'clean'
Data Padding: None
Keyword Arguments Passed to load(): {}
Custom Functions: 0 applied

Local File Statistics
---------------------
Number of files: 95
Date Range: 18 January 2010 --- 13 December 2015

Loaded Data Statistics
----------------------
No loaded data.
 
 Custom Functions: 0 applied
 
 Local File Statistics
---------------------
Number of files: 95
Date Range: 18 January 2010 --- 13 December 2015
```

If you have custom functions, the printed summary gives you the hidden information: function name, pysat kind, arguments, and kwargs.

**Comment by rstoneback** [comment]
I spent a bit of time poking at custom myself. Thought I was doing well.

```
In [2]: test.custom                                                                                                                                                   
Out[2]: pysat.Custom().attach(<function _pass at 0x7ff82055d430>, 'modify').attach(<function _pass at 0x7ff8205979d0>, 'modify', args=['hi'], kwargs={'custom2': 'yo'})
```
However, turns out Python functions don't work with a repr reconstruction,
```
In [4]: _pass                                                                                                                                                         
Out[4]: <function __main__._pass(inst, arg1, custom2=3)>

In [5]: eval(repr(_pass))                                                                                                                                             
Traceback (most recent call last):

  File "<string>", line 1
    <function _pass at 0x7ff8205979d0>
    ^
SyntaxError: invalid syntax
```
doesn't work, so... 
Doesn't look possible to recreate custom from repr.

I also poked at recreating from the files repr. It works, sometimes... the issue stems from my use of a weak reference within files to the containing pysat.Instrument.

```
In [34]: pysat.Files(pysat.Instrument(platform='pysat', name='testing', sat_id='', clean_level='clean', pad=None, orbit_info={'index': None, 'kind': None, 'period': N
    ...: one}, **{'sim_multi_file_right': False, 'sim_multi_file_left': False, 'root_date': None, 'file_date_range': None, 'malformed_index': False, 'mangle_file_date
    ...: s': False}))                                                                                                                                                 
Out[34]: Files(Instrument(platform='pysat', name='testing', sat_id='', clean_level='clean', pad=None, orbit_info={'index': None, 'kind': None, 'period': None}, **{'sim_multi_file_right': False, 'sim_multi_file_left': False, 'root_date': None, 'file_date_range': None, 'malformed_index': False, 'mangle_file_dates': False}), manual_org=False, directory_format={platform}/{name}/{tag}, update_files=False, file_format=None, write_to_diskTrue, ignore_empty_files=False) -> 1096 Local files

In [35]: pysat.Files(pysat.Instrument(platform='pysat', name='testing', sat_id='', clean_level='clean', pad=None, orbit_info={'index': None, 'kind': None, 'period': N
    ...: one}, **{'sim_multi_file_right': False, 'sim_multi_file_left': False, 'root_date': None, 'file_date_range': None, 'malformed_index': False, 'mangle_file_date
    ...: s': False}))                                                                                                                                                 
Out[35]: ---------------------------------------------------------------------------


~/Code/pysat/pysat/_files.py in __repr__(self)
    184 
    185     def __repr__(self):
--> 186         out_str = "".join(["Files(", self._sat.__repr__(), ", manual_org=",
    187                            "{:}, directory_format=".format(self.manual_org),
    188                            self.directory_format, ", update_files=",

ReferenceError: weakly-referenced object no longer exists

```


For a bit I thought it might be good to ensure pysat could make it through a full recreation using repr. It may still be worth doing, though, given Python functions don't work, I don't think it will work in the end.

File: pysat/_instrument.py [hunk 4]
```diff
@@ -1053,77 +1051,83 @@ def _assign_funcs(self, by_name=False, inst_module=None):
             # Either flags are not specified, or this combo is not
             self._password_req = False
 
-    def __str__(self):
+    def __repr__(self):
+        # Print the basic Instrument properties
+        out_str = "".join(["Instrument(platform='", self.platform, "', name='",
+                           self.name, "', sat_id='", self.sat_id,
```
Comment (rstoneback) [inline pysat/_instrument.py:1057]:
```suggestion
        out_str = "".join(["pysat.Instrument(platform='", self.platform, 
                           "', name='", self.name, "', sat_id='", self.sat_id,
```

File: pysat/_files.py [hunk 5]
```diff
@@ -180,6 +182,32 @@ def __init__(self, sat, manual_org=False, directory_format=None,
                 # couldn't find stored info, load file list and then store
                 self.refresh()
 
+    def __repr__(self):
+        out_str = "".join(["Files(", self._sat.__repr__(), ", manual_org=",
+                           "{:}, directory_format=".format(self.manual_org),
+                           self.directory_format, ", update_files=",
```
Comment (rstoneback) [inline pysat/_files.py:188]: self.directory format needs to be printed as a string, aka the string itself needs internal quotes. the %r style

File: pysat/_files.py [hunk 6]
```diff
@@ -180,6 +182,32 @@ def __init__(self, sat, manual_org=False, directory_format=None,
                 # couldn't find stored info, load file list and then store
                 self.refresh()
 
+    def __repr__(self):
+        out_str = "".join(["Files(", self._sat.__repr__(), ", manual_org=",
```
Comment (rstoneback) [inline pysat/_files.py:186]: pysat.Files

File: pysat/tests/test_instrument.py [hunk 7]
```diff
@@ -29,25 +29,27 @@ def setup(self):
...
         """Test if the correct day is being loaded (checking object date and
         data)."""
-        self.testInst.load(2009, 1)
-        test_date = self.testInst.index[0]
-        test_date = dt.datetime(test_date.year, test_date.month,
-                                test_date.day)
-        assert (test_date == dt.datetime(2009, 1, 1))
-        assert (test_date == self.testInst.date)
+        self.testInst.load(self.ref_time.year, self.ref_doy)
+        self.out = self.testInst.index[0]
```
Comment (rstoneback) [inline pysat/tests/test_instrument.py:49]: Missed all these changes the first time :P

Can you provide some background on why internal variables within the test functions are being assigned to the test object?

File: pysat/_instrument.py (hunk 4 above)
Comment (aburrell) [reply pysat/_instrument.py:1057]: I don't agree with this suggestion.  Following the example of numpy, once you have declared a class object such as an array, the __repr__ looks like: `array('string', dtype='<U6')` and no reference to the path to get to the class is made.

File: pysat/_files.py (hunk 6 above)
Comment (aburrell) [reply pysat/_files.py:186]: See previous comment on Instrument.

File: pysat/tests/test_instrument.py (hunk 7 above)
Comment (aburrell) [reply pysat/tests/test_instrument.py:49]: No, you just missed the commit history :P 

This is reducing redundancy in the test code and improving its maintainability.  When you do this, you also can clearly see how many of the tests are duplicates of each other and require parametrization, but I opened an issue for that already.

File: pysat/_files.py (hunk 5 above)
Comment (aburrell) [reply pysat/_files.py]: Fixed in latest commit.

File: pysat/tests/test_custom.py [hunk 8]
```diff
@@ -43,11 +43,42 @@ def setup(self):
...
+    def test_basic_str(self):
+        """Check for lines from each decision point in str"""
+        self.out = self.testInst.custom.__str__()
+        assert isinstance(self.out, str)
+        # No custom functions
+        assert self.out.find('0 applied') > 0
+
+    def test_basic_str_w_function(self):
+        """Check for lines from each decision point in str"""
+        def mult_data(inst, mult, dkey="mlt"):
```
Comment (rstoneback) [inline pysat/tests/test_custom.py:68]: The custom mult_data function isn't being run.

File: pysat/_instrument.py [hunk 9]
```diff
@@ -1053,77 +1051,83 @@ def _assign_funcs(self, by_name=False, inst_module=None):
...
+            output_str += 'Number of Times: {:d}\n'.format(len(self.index))
+            output_str += 'Number of variables: {:d}\n'.format(num_vars)
+
+            if num_vars <= max_vars:
+                output_str += '\nVariable Names:\n'
+                num = len(self.variables) // 3
+                for i in np.arange(num):
+                    output_str += self.variables[3 * i].ljust(30)
+                    output_str += self.variables[3 * i + 1].ljust(30)
+                    output_str += self.variables[3 * i + 2].ljust(30) + '\n'
```
Comment (rstoneback) [inline pysat/_instrument.py:1122]: These few output string lines are not being tested.

File: pysat/_instrument.py (hunk 4 above)
Comment (rstoneback) [reply pysat/_instrument.py:1057]: You are right about the numpy standard. I checked a few others and results are mixed:
```
In [14]: np.array([0, 1, 2]).__repr__()                                                                                                                               
Out[14]: 'array([0, 1, 2])'

In [8]: dt.datetime(2009, 1, 1).__repr__()                                                                                                                            
Out[8]: 'datetime.datetime(2009, 1, 1, 0, 0)'

In [11]: xarray.Dataset().__repr__()                                                                                                                                  
Out[11]: '<xarray.Dataset>\nDimensions:  ()\nData variables:\n    *empty*'

In [13]: pds.DataFrame().__repr__()                                                                                                                                   
Out[13]: 'Empty DataFrame\nColumns: []\nIndex: []'

```
I do still prefer including the pysat but I am ok with leaving it off.

File: pysat/tests/test_custom.py (hunk 8 above)
Comment (aburrell) [reply pysat/tests/test_custom.py:68]: Yup, it doesn't need to be run.  It just needs to be assigned.  This could be cleaned up by defining this function in setup or outside the Class since there's lots of duplication of defining small custom routines in this test suite.  However, that's a separate issue that should be created.

File: pysat/_instrument.py (hunk 9 above)
Comment (aburrell) [reply pysat/_instrument.py:1122]: And I have no idea why, because I took great pains to make a test that does print out data variables.  Can you try out `test_instrument.py::TestBasics::test_str_w_load_less_data` and see WHY it hates those lines?

**Comment by aburrell** [comment]
> There is a remaining gap for Files on the weakref. There may be a test by repeatedly doing the eval(repr(inst.files)) until it triggers.

If you want to mock up the test for that, or just submit it as an issue, we can ensure your solution isn't lost.  Any idea how many times you have to do eval before the weakref breaks?

**Comment by rstoneback** [comment]
> > There is a remaining gap for Files on the weakref. There may be a test by repeatedly doing the eval(repr(inst.files)) until it triggers.
> 
> If you want to mock up the test for that, or just submit it as an issue, we can ensure your solution isn't lost. Any idea how many times you have to do eval before the weakref breaks?

Mostly it didn't work. It did work once though, so I know it can.

**Comment by rstoneback** [comment]
> > > There is a remaining gap for Files on the weakref. There may be a test by repeatedly doing the eval(repr(inst.files)) until it triggers.
> > 
> > 
> > If you want to mock up the test for that, or just submit it as an issue, we can ensure your solution isn't lost. Any idea how many times you have to do eval before the weakref breaks?
> 
> Mostly it didn't work. It did work once though, so I know it can.

Created issue #534. 42

File: pysat/tests/test_instrument.py [hunk 10]
```diff
@@ -382,42 +394,58 @@ def test_repr_w_orbit(self):
...
+        assert self.out.find('Number of variables:') > 0
+        assert self.out.find('uts') < 0
+
+    def test_str_w_load_less_data(self):
+        """Test string output with loaded data """
+        # Load the test data
+        self.testInst.load(self.ref_time.year, self.ref_doy)
+
+        # Ensure the desired data variable is present and delete all others
+        self.testInst.data = self.testInst.data[['mlt']]
```
Comment (rstoneback) [inline pysat/tests/test_instrument.py:436]:
```suggestion
        # minimum of 4 variables needed to test all lines
        self.testInst.data = self.testInst.data[self.testInst.variables[0:4]]
```

File: pysat/_instrument.py (hunk 9 above)
Comment (rstoneback) [reply pysat/_instrument.py:1122]: Suggestion below.

File: pysat/tests/test_instrument.py (hunk 10 above)
Comment (aburrell) [reply pysat/tests/test_instrument.py:436]: Thanks for figuring out the reason why this was going wrong!  The suggested change isn't sufficient to fix the test, so I committed something more complete below.

**Review by rstoneback — APPROVED** [review]
Thanks @aburrell! I was trying to use these the other day on a new instrument and was not very happy with it. Then shortly after your pull popped up. Nice timing. 

The doc pull is still waiting for an update from Matthew. He tried to alter the current path for python 3 and wasn't able to make it work. He is trying to use WSL instead.

We will need to give Meta a similar repr refresh when it goes through its review.
